File: macos/DotsHarness/Sources/DotsHarnessCore/BrowserContracts.swift
// Copyright (c) 2026 DOTS
// Browser capability contracts for the native macOS harness.

import Foundation

public enum BrowserBackend: String, Codable, Sendable, Equatable {
    case managed = "Managed"
    case `extension` = "Extension"
    case unknown = "Unknown"
}

public enum BrowserCleanupStatus: String, Codable, Sendable, Equatable {
    case notUsed = "NotUsed"
    case closed = "Closed"
    case pendingRecovery = "PendingRecovery"
    case failed = "Failed"
}

public struct BrowserScope: Codable, Sendable, Equatable, Hashable {
    public let area: String
    public let conversationID: String
    public let runID: String

    public init(area: String, conversationID: String, runID: String) {
        self.area = area
        self.conversationID = conversationID
        self.runID = runID
    }

    public var key: String { "\(area):\(conversationID):\(runID)" }
}

public struct BrowserPage: Codable, Sendable, Equatable {
    public let pageID: String
    public let backend: BrowserBackend
    public let url: String
    public let scope: BrowserScope

    public init(pageID: String, backend: BrowserBackend, url: String, scope: BrowserScope) {
        self.pageID = pageID
        self.backend = backend
        self.url = url
        self.scope = scope
    }
}

public struct BrowserRunSummary: Codable, Sendable, Equatable, Hashable {
    public let backend: BrowserBackend
    public let openedPageIDs: [String]
    public let closedPageIDs: [String]
    public let cleanupStatus: BrowserCleanupStatus
    public let diagnosticCode: String

    public init(
        backend: BrowserBackend,
        openedPageIDs: [String] = [],
        closedPageIDs: [String] = [],
        cleanupStatus: BrowserCleanupStatus = .notUsed,
        diagnosticCode: String = "not_used"
    ) {
        self.backend = backend
        self.openedPageIDs = openedPageIDs
        self.closedPageIDs = closedPageIDs
        self.cleanupStatus = cleanupStatus
        self.diagnosticCode = diagnosticCode
    }
}

public enum BrowserBackendPolicy {
    public static func resolveUserRule(_ userText: String?) -> BrowserBackend {
        guard let userText, !userText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return .unknown
        }
        let value = userText.lowercased()
        if containsAny(value,
                       "mevcut chrome", "chrome profilimi", "chrome oturumumu", "oturum açılmış chrome",
                       "oturum acilmis chrome", "current chrome", "existing chrome", "logged-in chrome",
                       "logged in chrome", "use my chrome", "use existing chrome", "my chrome profile") {
            return .`extension`
        }
        if containsAny(value,
                       "managed browser", "isolated browser", "clean browser", "temporary browser",
                       "izole tarayıcı", "izole tarayici", "temiz tarayıcı", "temiz tarayici",
                       "managed chrome", "clean profile", "temiz profil") {
            return .managed
        }
        return .unknown
    }

    public static func parseConfirmation(_ answer: String?) -> BrowserBackend {
        guard let answer, !answer.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return .unknown
        }
        let value = answer.lowercased()
        if containsAny(value, "extension", "mevcut chrome", "current chrome", "existing chrome", "chrome profile", "existing chrome profile")
            || containsToken(value, "2") {
            return .`extension`
        }
        if containsAny(value, "managed", "izole", "isolated", "clean", "temiz", "managed isolated browser")
            || containsToken(value, "1") {
            return .managed
        }
        return .unknown
    }

    private static func containsAny(_ value: String, _ candidates: String...) -> Bool {
        candidates.contains { value.contains($0) }
    }

    private static func containsToken(_ value: String, _ token: String) -> Bool {
        value.split { character in
            character.isWhitespace || ".,;:!?()[]{}\"'-".contains(character)
        }.contains { $0 == token }
    }
}

public enum BrowserTools {
    public static let openName = "browser_open"
    public static let navigateName = "browser_navigate"
    public static let closeName = "browser_close"

    public static let definitions: [AgentToolDefinition] = [
        AgentToolDefinition(
            name: openName,
            description: "Open a URL in the host-managed browser session. The host selects the backend; do not pass a backend.",
            parameters: .object([
                "type": .string("object"),
                "properties": .object([
                    "url": .object([
                        "type": .string("string"),
                        "description": .string("An http or https URL."),
                    ]),
                ]),
                "required": .array([.string("url")]),
                "additionalProperties": .bool(false),
            ])
        ),
        AgentToolDefinition(
            name: navigateName,
            description: "Navigate an owned browser page. The pageID must come from browser_open.",
            parameters: .object([
                "type": .string("object"),
                "properties": .object([
                    "pageID": .object(["type": .string("string")]),
                    "url": .object([
                        "type": .string("string"),
                        "description": .string("An http or https URL."),
                    ]),
                ]),
                "required": .array([.string("pageID"), .string("url")]),
                "additionalProperties": .bool(false),
            ])
        ),
        AgentToolDefinition(
            name: closeName,
            description: "Close an owned browser page. Pages are also closed automatically when the run ends.",
            parameters: .object([
                "type": .string("object"),
                "properties": .object([
                    "pageID": .object(["type": .string("string")]),
                ]),
                "required": .array([.string("pageID")]),
                "additionalProperties": .bool(false),
            ])
        ),
    ]

    public static func backendQuestionCall(_ callID: String) -> AgentToolCall {
        AgentToolCall(
            id: "\(callID):browser-backend",
            name: AskUserTool.name,
            arguments: #"{"questions":[{"header":"Browser","question":"Which browser session should this run use?","options":["Managed isolated browser","Existing Chrome profile"]}]}"#
        )
    }
}

File: shared/BrowserContracts.cs
// Copyright (c) 2026 DOTS
// Browser capability contracts shared by the Windows and Linux harnesses.

using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace DotsHarnessCore;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum BrowserBackend
{
    Managed,
    Extension,
    Unknown,
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum BrowserCleanupStatus
{
    NotUsed,
    Closed,
    PendingRecovery,
    Failed,
}

public readonly record struct BrowserScope(
    [property: JsonPropertyName("area")] string Area,
    [property: JsonPropertyName("conversationID")] string ConversationId,
    [property: JsonPropertyName("runID")] string RunId)
{
    public string Key => $"{Area}:{ConversationId}:{RunId}";
}

public sealed record BrowserPage(
    [property: JsonPropertyName("pageID")] string PageId,
    [property: JsonPropertyName("backend")] BrowserBackend Backend,
    [property: JsonPropertyName("url")] string Url,
    [property: JsonPropertyName("scope")] BrowserScope Scope);

public sealed record BrowserRunSummary(
    [property: JsonPropertyName("backend")] BrowserBackend Backend,
    [property: JsonPropertyName("openedPageIDs")] IReadOnlyList<string> OpenedPageIds,
    [property: JsonPropertyName("closedPageIDs")] IReadOnlyList<string> ClosedPageIds,
    [property: JsonPropertyName("cleanupStatus")] BrowserCleanupStatus CleanupStatus,
    [property: JsonPropertyName("diagnosticCode")] string DiagnosticCode)
{
    public static BrowserRunSummary NotUsed(BrowserBackend backend = BrowserBackend.Unknown) =>
        new(backend, Array.Empty<string>(), Array.Empty<string>(), BrowserCleanupStatus.NotUsed, "not_used");
}

public enum AgentCommandSandboxMode
{
    Legacy,
    StrictNoDesktop,
}

public static class BrowserBackendPolicy
{
    public static BrowserBackend ResolveUserRule(string? userText)
    {
        if (string.IsNullOrWhiteSpace(userText)) return BrowserBackend.Unknown;
        var value = userText.Trim().ToLowerInvariant();

        if (ContainsAny(value,
                "mevcut chrome", "chrome profilimi", "chrome oturumumu", "oturum açılmış chrome",
                "oturum acilmis chrome", "current chrome", "existing chrome", "logged-in chrome",
                "logged in chrome", "use my chrome", "use existing chrome", "my chrome profile"))
            return BrowserBackend.Extension;

        if (ContainsAny(value,
                "managed browser", "isolated browser", "clean browser", "temporary browser",
                "izole tarayıcı", "izole tarayici", "temiz tarayıcı", "temiz tarayici",
                "managed chrome", "clean profile", "temiz profil"))
            return BrowserBackend.Managed;

        return BrowserBackend.Unknown;
    }

    public static BrowserBackend ParseConfirmation(string? answer)
    {
        if (string.IsNullOrWhiteSpace(answer)) return BrowserBackend.Unknown;
        var value = answer.Trim().ToLowerInvariant();
        if (ContainsAny(value, "extension", "mevcut chrome", "current chrome", "existing chrome", "chrome profile", "existing chrome profile")
            || ContainsToken(value, "2"))
            return BrowserBackend.Extension;
        if (ContainsAny(value, "managed", "izole", "isolated", "clean", "temiz", "managed isolated browser")
            || ContainsToken(value, "1"))
            return BrowserBackend.Managed;
        return BrowserBackend.Unknown;
    }

    private static bool ContainsAny(string value, params string[] candidates) =>
        candidates.Any(candidate => value.Contains(candidate, StringComparison.Ordinal));

    private static bool ContainsToken(string value, string token) =>
        value.Split([' ', '\t', '\r', '\n', '.', ',', ':', ';', '!', '?', '(', ')', '[', ']', '{', '}', '"', '\'', '-'],
                StringSplitOptions.RemoveEmptyEntries)
            .Any(part => string.Equals(part, token, StringComparison.Ordinal));
}

public static class BrowserTools
{
    public const string OpenName = "browser_open";
    public const string NavigateName = "browser_navigate";
    public const string CloseName = "browser_close";

    public static IReadOnlyList<NativeToolDefinition> Definitions { get; } =
    [
        new(OpenName,
            "Open a URL in the host-managed browser session. The host selects the backend; do not pass a backend.",
            new JsonObject
            {
                ["type"] = "object",
                ["properties"] = new JsonObject
                {
                    ["url"] = new JsonObject { ["type"] = "string", ["description"] = "An http or https URL." },
                },
                ["required"] = new JsonArray("url"),
                ["additionalProperties"] = false,
            }),
        new(NavigateName,
            "Navigate an owned browser page. The pageID must come from browser_open.",
            new JsonObject
            {
                ["type"] = "object",
                ["properties"] = new JsonObject
                {
                    ["pageID"] = new JsonObject { ["type"] = "string" },
                    ["url"] = new JsonObject { ["type"] = "string", ["description"] = "An http or https URL." },
                },
                ["required"] = new JsonArray("pageID", "url"),
                ["additionalProperties"] = false,
            }),
        new(CloseName,
            "Close an owned browser page. Pages are also closed automatically when the run ends.",
            new JsonObject
            {
                ["type"] = "object",
                ["properties"] = new JsonObject
                {
                    ["pageID"] = new JsonObject { ["type"] = "string" },
                },
                ["required"] = new JsonArray("pageID"),
                ["additionalProperties"] = false,
            }),
    ];

    public static NativeToolCall BackendQuestionCall(string callId) => new(
        $"{callId}:browser-backend",
        AskUserTool.Name,
        """{"questions":[{"header":"Browser","question":"Which browser session should this run use?","options":["Managed isolated browser","Existing Chrome profile"]}]}""");
}
